File: 3.breath-first-search/동규/2.perfect-squares.js
// dp(n) = min(dp(n-X)+1)  => X는 n을 넘기지 않는 완전 제곱수
// dp(13)은 9를 완전제곱수로 선택 후 dp(13-9)+1 과 같이 표현

function numSquares(n) {
  const dp = new Array(n + 1).fill(Infinity);
  dp[0] = 0;

  for (let i = 1; i <= n; i++) {
    let p = 1;
    while (p ** 2 <= i) {
      dp[i] = Math.min(dp[i], dp[i - p ** 2] + 1);
      p += 1;
    }
  }
  return dp[n];
}

File: 3.breath-first-search/연경/2.perfect-squares.py
import math
import sys


class Solution:
    def numSquares(self, n: int) -> int:

        dp = [sys.maxsize for _ in range(n + 1)]
        dp[0] = 0
        dp[1] = 1

        if n <= 1:
            return dp[n]
        # 1 + 1
        dp[2] = 2
        # # 1 + 1 + 1
        # dp[3] = 3
        # dp[4] = 1
        # # 1 + 4
        # dp[5] = 2
        for i in range(3, n + 1):
            for j in range(1, int(math.sqrt(i)) + 1):
                dp[i] = min(dp[i], dp[i - j * j] + 1)
        return dp[n]

if __name__ == "__main__":
    n = 6
    print(Solution().numSquares(n))